### PR TITLE
refactor(processor): add range check trace table

### LINKED ISSUE #229: Optimize range checker execution trace finalization / length check
Currently, the 8-bit range check table is built twice when the execution trace is finalized.

1. It is built during the [call to `RangeChecker::trace_len`](https://github.com/maticnetwork/miden/blob/78f4b7d228aca6d0949ac37dc9854d6662a2ee8a/processor/src/trace.rs#L222) when the maximum length of the execution trace is being computed in `finalize_trace`.
2. It is built during the [`range.into_trace` call](https://github.com/maticnetwork/miden/blob/78f4b7d228aca6d0949ac37dc9854d6662a2ee8a/processor/src/trace.rs#L240) before the range checker's trace is combined into the final execution trace.

Ideally, it would be better to build this table just once. It is probably also worth checking the finalization of other processors to ensure there are no similar duplicated computations.

### PR DESCRIPTION
**Pull request by vlopes11**
## Describe your changes

This commit introduces `RangeCheckTraceTable`, a struct that will contain the 8bit lookups table for the range trace, and the combined length of the 8bits and 16bits tables.

Prior to this commit, the table was generated twice - first to calculate the `max_len`, and then to compute the trace.

From local benchmarks, this optimization improved the performance around ~3%, considering we no longer allocate the tables twice. However, the performance improvement could be mitigated by the compiler since he should be smart enough to understand that the table shouldn't be dropped before the second call to `into_trace`.

resolves #229

## Checklist before requesting a review
- [x] Repo forked and branch created from `next` according to naming convention.
- [x] Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- [x] Relevant issues are linked in the PR description.
- [ ] Tests added for new functionality. (N/A, refactor without behavior change)
- [ ] Documentation/comments updated according to changes. (N/A, internal refactor)

### CODE REVIEW

File: core/src/lib.rs
Comment (grjte) [inline core/src/lib.rs:83]: For consistency, I don't think this is the best place for this constant. The top level here contains stuff that is a constant describing the shape of the trace (main or auxiliary) or globally used. In general, we have created specific modules within this crate for each VM component that has constants which are used in multiple crates (e.g. `decoder`, `stack`, etc). However, when constants aren't needed in multiple crates, we tend to just define them in the module/crate where they're needed. 

In this case, this constant should probably be defined within the `range` module in the `processor` crate

File: processor/src/lib.rs
Comment (grjte) [inline processor/src/lib.rs:78]: Similar comment here - I would keep this in the `range` module since it's not actually needed at this level.

File: processor/src/range/mod.rs
Comment (grjte) [inline processor/src/range/mod.rs:86]: nit: this seems to only be used in 1 test  now, so maybe we should mark it `#[cfg(test)]` and move it down down to the bottom under a "TEST HELPERS" section? Or at least grouped with the other public accessors?

File: processor/src/range/mod.rs
Comment (grjte) [inline processor/src/range/mod.rs:138]: similar nit: this function is only used for tests, so maybe we should mark it `#[cfg(test)]` and group it with the other one at the bottom? (I think it would just give a bit more clarity on when the 8bit lookup table is actually being built when the VM is executed)

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @vlopes11! I really like this approach & how clean it is. I left a couple of small things inline related to code organization

File: processor/src/range/mod.rs
Comment (vlopes11) [reply processor/src/range/mod.rs:138]: This function could instead be removed. I was a bit concerned to cause a breaking change so that's why I opted to keep it. But if breaking the API wouldn't be a problem here, its better we just centralize the usage into a single `into_trace` that will take the tables as arguments. wdyt?

Comment (grjte) [reply processor/src/range/mod.rs:138]: Ah, great point about the API. Actually I think all of these `into_trace` methods could probably be marked `pub(super)` or `pub(crate)`. They're not called outside the `processor`, and I don't see a reason why they should be. Am I missing something @bobbinth?

1. Yes, I think we could remove this wrapper and update the tests to use the method that takes the table.
2. I actually do like the current naming, since all of the other `into_trace` methods have the same signature (which doesn't take a table), but this is just a very small preference, so I think whatever you want to do with the naming is fine.

Comment (vlopes11) [reply processor/src/range/mod.rs:138]: To add to the context before @bobbinth replies; I agree with `2`. Having a different name will differentiate from the other variants that won't take a table and this will create an unmistakable domain separation (just repeating what you said haha)

Comment (bobbinth) [reply processor/src/range/mod.rs:138]: Sorry for the delay! I agree with what you guys have discussed:

1. Let's keep the name `into_trace_with_table`.
2. Let's remove `into_trace` wrapper and update the tests accordingly (if needed, we could add a helper function there).
3. I'm not sure about setting visibility to `pub(crate)` or `pub(super)`. If we wanted to keep things consistent, we'd have to apply the same to many other methods as well. Also, if I'm not mistaken, none of these components are actually exported from the crate - so, they are not a part of the crate's public interface anyway.

File: core/src/lib.rs
Comment (vlopes11) [reply core/src/lib.rs]: Done!

File: processor/src/lib.rs
Comment (vlopes11) [reply processor/src/lib.rs]: Done!

File: processor/src/range/mod.rs
Comment (vlopes11) [reply processor/src/range/mod.rs]: Done!

File: processor/src/range/mod.rs
Comment (vlopes11) [reply processor/src/range/mod.rs]: Confirmed, these are not exported. I kept them as public, so we do a more general refactor in the future if we want to introduce `pub(crate)`

**Review by bobbinth — APPROVED** [review]
All looks good to me! Thank you!

File: processor/src/range/mod.rs
Comment (grjte) [inline processor/src/range/mod.rs:29]: nit: is `len` the 16-bit length or is it the length of 8-bit plus 16-bit? (It looks like the latter)

**Review by grjte — APPROVED** [review]
Thanks @vlopes11! It looks great. I left one nit inline about the docs, but I'll approve. If you agree with the nit, you can update that change before merging.

When you merge, please also update the tracking PR #361 to mark this done ("Optimize range checker execution trace finalization.")

File: processor/src/range/mod.rs
Comment (vlopes11) [reply processor/src/range/mod.rs]: Its confusing indeed; updated. Its the combined length of 8 and 16 bits

**Comment by vlopes11** [comment]
Blocked by https://github.com/maticnetwork/miden/pull/447

**Review by bobbinth — APPROVED** [review]
All looks good! Once #447 is merged (in a few mins) - we can merge this one too.

**Pull request merged**
